gfx.cpp

```diff
@@ -1502,84 +1527,162 @@ class GfxInternal
         gfxFree(fence_values_);
+        fence_values_ = nullptr;
         gfxFree(fences_);
+        fences_ = nullptr;
 
         if(dxc_utils_ != nullptr)
+        {
             dxc_utils_->Release();
+            dxc_utils_ = nullptr;
+        }
         if(dxc_compiler_ != nullptr)
+        {
             dxc_compiler_->Release();
+            dxc_compiler_ = nullptr;
+        }
         if(dxc_include_handler_ != nullptr)
+        {
             dxc_include_handler_->Release();
+            dxc_include_handler_ = nullptr;
+        }
 
         if(swap_chain_ != nullptr)
-            swap_chain_->Release();
-        else
         {
-            if(back_buffer_allocations_ != nullptr)
-                for(uint32_t i = 0; i < max_frames_in_flight_; ++i)
-                    if(back_buffer_allocations_[i] != nullptr)
-                        back_buffer_allocations_[i]->Release();
-            gfxFree(back_buffer_allocations_);
+            swap_chain_->Release();
+            swap_chain_ = nullptr;
         }
+        if(back_buffer_allocations_ != nullptr)
+            for(uint32_t i = 0; i < max_frames_in_flight_; ++i)
+                if(back_buffer_allocations_[i] != nullptr)
+                    back_buffer_allocations_[i]->Release();
+        gfxFree(back_buffer_allocations_);
+        back_buffer_allocations_ = nullptr;
         if(back_buffers_ != nullptr)
             for(uint32_t i = 0; i < max_frames_in_flight_; ++i)
                 if(back_buffers_[i] != nullptr)
                     back_buffers_[i]->Release();
         gfxFree(back_buffer_rtvs_);
+        back_buffer_rtvs_ = nullptr;
         gfxFree(back_buffers_);
+        back_buffers_ = nullptr;
 
         if(mem_allocator_ != nullptr)
+        {
             mem_allocator_->Release();
+            mem_allocator_ = nullptr;
+        }
         if(dispatch_signature_ != nullptr)
+        {
             dispatch_signature_->Release();
+            dispatch_signature_ = nullptr;
+        }
         if(multi_draw_signature_ != nullptr)
+        {
             multi_draw_signature_->Release();
+            multi_draw_signature_ = nullptr;
+        }
         if(multi_draw_indexed_signature_ != nullptr)
+        {
             multi_draw_indexed_signature_->Release();
+            multi_draw_indexed_signature_ = nullptr;
+        }
         if(dispatch_rays_signature_ != nullptr)
+        {
             dispatch_rays_signature_->Release();
+            dispatch_rays_signature_ = nullptr;
+        }
         if(draw_mesh_signature_ != nullptr)
+        {
             draw_mesh_signature_->Release();
+            draw_mesh_signature_ = nullptr;
+        }
+
+#ifdef PIX_AMD_EXT
+        tls_pAmdExtDeviceObject = nullptr;
+        tls_checkAmdDriver = true;
+#endif
 
         return kGfxResult_NoError;
     }
 
+    bool isValid() const
+    {
+        return device_ != nullptr && mem_allocator_ != nullptr && device_->GetDeviceRemovedReason() == S_OK;
+    }
```

Comment:
Added a new function that allows to check if a gfx context has been correctly initialized and that it is still valid (i.e. device is still valid)